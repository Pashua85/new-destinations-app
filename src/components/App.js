import React from 'react';
import uuidv1 from 'uuid/v1';

import Header from './Header';
import MapComponent from './MapComponent';
import AddPoint from './AddPoint';
import PointList from './PointList';
import ErrorMessage from './ErrorMessage';

class App extends React.Component {
  state = {
    points: [],
    pathCoords: [],
    currentCenter: {
      lat: 55.751244,
      lng: 37.618423
    },
    errorMessage: ''  
  };

  setCurrentCenter = (currentCenter) => {
    this.setState({ currentCenter });
    console.log(this.state.currentCenter);
  };

  isUniquePlace = (newPoint) => {
    if(this.state.points.length > 0) {
      const repeatingPoints = this.state.points.filter(point => {
        return point.latLng.lat === newPoint.latLng.lat 
          && point.latLng.lng === newPoint.latLng.lng;
      });
      return repeatingPoints.length === 0;
    }
    return true;
  }; 

  addPoint = (pointName) => {
    const newPoint = {
      pointName: pointName,
      pointId: uuidv1(),
      latLng: this.state.currentCenter
    };
    if (this.isUniquePlace(newPoint)) {
      this.setState((prevState) => (
        {
          points: prevState.points.concat(newPoint),
          errorMessage: ''
        }
      ));
      this.setPathCoords();
    } else {
      this.handleError('There is a point on this place already')
    }
  };

  handleError = (errorMessage) => {
    this.setState({ errorMessage });
  };

  reorderPoints = (points) => {
    this.setState(() => ({ points }));
    this.setPathCoords();
  };

  deletePoint = (pointId) => {
    this.setState(prevState => ({
      points: prevState.points.filter(point => point.pointId !== pointId)
    }));
    this.setPathCoords();
  };

  setPathCoords = () => {
    this.setState((prevState) => ({
      pathCoords: prevState.points.map(point => point.latLng)
    }));
  };

  onChangePosition = (pointId, newLatLng) => {
    const pointsCopy = this.state.points.slice();
    const newPoints = this.state.points.map( point => {
      if (point.pointId === pointId) {
        point.latLng = newLatLng;
        return point;
      } else {
        return point;
      }
    });
    this.setState(() => ({
      points: newPoints
    }));
    this.setPathCoords();
  };

  render() {
    return (
      <div className="app-container">
        <Header />
        <div className="main-container container">
          <div className="widget">
            <AddPoint 
              handleAddPoint={this.addPoint}
              handleError={this.handleError}
            />
            <ErrorMessage errorMessage={this.state.errorMessage} />
            <PointList
              handleReorder={this.reorderPoints} 
              points={this.state.points}
              deletePoint={this.deletePoint}  
            /> 
          </div>
          <MapComponent
            isMarkerShown points={this.state.points} 
            onChangePosition={this.onChangePosition}
            pathCoords={this.state.pathCoords}
            onCenterChanged={this.setCurrentCenter}
          />
        </div>
      </div>
    )
  }
};

export default App;