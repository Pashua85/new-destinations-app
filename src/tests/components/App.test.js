import React from 'react';
import { shallow } from 'enzyme';
import App from '../../components/App';
import points from '../utils/points';
import pathCoords from '../utils/pathCoords';

let wrapper;
const pointName = 'test name';
const currentCenter = {
  lat: 55.751244,
  lng: 37.618423
}

beforeEach(() => {
  wrapper = shallow(<App />);
  wrapper.find('AddPoint').prop('handleAddPoint')(pointName);
})

test('should render App', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should add points', () => {
  expect(wrapper.state('points').length).toEqual(1);
});

test('should set pathCoords', () => {
  expect(wrapper.state('pathCoords')[0]).toEqual(currentCenter);
});

test('should delete address', () => {
  const placeIdToDelete = wrapper.state('points')[0].pointId;
  wrapper.find('PointList').prop('deletePoint')(placeIdToDelete);
  expect(wrapper.state('points').length).toEqual(0);
});

test('should set error message and render it correctly', () => {
  wrapper.find('AddPoint').prop('handleError')('Test error message');
  expect(wrapper.state('errorMessage')).toBe('Test error message');
  expect(wrapper).toMatchSnapshot();
});

test('should throw an error for trying to add points on the same place', () => {
  wrapper.find('AddPoint').prop('handleAddPoint')('another test name');
  expect(wrapper.state('points').length).toEqual(1);
  expect(wrapper.state('errorMessage')).toEqual('There is a point on this place already');
});