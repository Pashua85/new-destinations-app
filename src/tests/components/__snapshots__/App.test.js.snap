// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render App 1`] = `
<div
  className="app-container"
>
  <Header />
  <div
    className="main-container container"
  >
    <div
      className="widget"
    >
      <AddPoint
        handleAddPoint={[Function]}
        handleError={[Function]}
      />
      <ErrorMessage
        errorMessage=""
      />
      <PointList
        deletePoint={[Function]}
        handleReorder={[Function]}
        points={Array []}
      />
    </div>
    <withProps(withGoogleMap(Map))
      isMarkerShown={true}
      onCenterChanged={[Function]}
      onChangePosition={[Function]}
      pathCoords={Array []}
      points={Array []}
    />
  </div>
</div>
`;

exports[`should set error message and render it correctly 1`] = `
<div
  className="app-container"
>
  <Header />
  <div
    className="main-container container"
  >
    <div
      className="widget"
    >
      <AddPoint
        handleAddPoint={[Function]}
        handleError={[Function]}
      />
      <ErrorMessage
        errorMessage=""
      />
      <PointList
        deletePoint={[Function]}
        handleReorder={[Function]}
        points={Array []}
      />
    </div>
    <withProps(withGoogleMap(Map))
      isMarkerShown={true}
      onCenterChanged={[Function]}
      onChangePosition={[Function]}
      pathCoords={Array []}
      points={Array []}
    />
  </div>
</div>
`;
